docs/testing.md

```diff
@@ -0,0 +1,179 @@
+
+## Unit tests
+
+The lowest level at which you can test is unit testing. Unit tests cover the smallest parts of code, like individual functions or classes.
+
+When the object under test has any dependencies, you’ll often mock them out, as described in the next paragraph.
+
+The great thing about unit tests is that they are quick to write and run. Therefore, as you work, you get fast feedback about whether your tests are passing. Jest even has an option to continuously run tests that are related to code you’re editing: [Watch mode](https://jestjs.io/docs/en/cli#watch).
+
+### Mocking
+
+Sometimes, when your tested objects have external dependencies, you’ll want to mock them out. “Mocking” is when you replace some dependency of your code by your own implementation. Note that generally, using real objects in your tests is better than using mocks but there are situations where this is not possible (for example when your JS code relies on a native module written in Java or Objective-C).
+
+Imagine you’re writing an app that shows the current weather in your city and you’re using some external service (dependency of your code) that provides you with the weather information. If the service tells you that it’s raining, you want to show an image with a rainy cloud. You don’t want to call that service in your tests, because it would make the tests slow and unstable (because of the network requests involved) and because the service may return different data every time you run the test. Therefore, you can provide a mock implementation of the service - effectively replacing thousands of lines of code and some internet-connected thermometers!
+
+Jest comes with [support for mocking](https://jestjs.io/docs/en/mock-functions#mocking-modules) from function level all the way to module level mocking.
+
+## Integration Tests
+
+When writing larger software systems, individual pieces of it need to interact with each other. As explained, in unit testing, when your unit depends on another one, you’ll sometimes end up mocking the dependency, replacing it with a fake one.
+
+In integration testing, real individual units are combined (same as in your app) and tested together to ensure that their cooperation works as expected. This is not to say that mocking does not happen here: you’ll still need mocks (for example to mock communication with the weather service covered in the mocking paragraph), but much less than in unit testing.
+
+> Please note that the terminology around what integration testing means is not always consistent. Also, the line between what is a unit test and what is an integration test may not always be clear. As a rule of thumb for this document, when your test
+>
+> - Uses an external system (such as the thermometer in the mocking paragraph)
+> - Does a network call to other application
+> - Does any kind of file or database I/O
+>
+> Then it falls into the “integration test” category.
+
+## Component Tests
```

Comment:
Is this supposed to be a sub point of Integration Tests? It is at the same heading level

Reply:
I'm myself not sure. I think component tests can fall below unit tests as well as integration tests, so I covered it separately, also because it's a little different (requires extra deps) from testing vanilla js.

Reply:
I addressed this by saying "Component tests could fall into both unit and integration testing, and since they are such a core part of React Native, we'll cover them in separate paragraphs."